pkg/hubble/observer/utils.go

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Hubble
+
+package observer
+
+import (
+	"fmt"
+	"strings"
+
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+)
+
+type filter map[string]filter
+
+// createFilter constructs a tree filter based on validated and normalized field
+// mask fm. Use active() to check if applying a filter will have any effect.
+func createFilter(fm *fieldmaskpb.FieldMask) (filter, error) {
+	if fm == nil {
+		return nil, nil
+	}
+	if !fm.IsValid(&flowpb.Flow{}) {
+		return nil, fmt.Errorf("invalid fieldmask")
+	}
+	fm.Normalize()
+
+	f := make(filter)
+	for _, path := range fm.GetPaths() {
+		f.add(path)
+	}
+	return f, nil
+}
+
+// add recursively converts path into tree based on its dot separated
+// components.
+func (f filter) add(path string) {
+	sub := strings.SplitN(path, ".", 2)
+	switch len(sub) {
+	case 1:
+		// ["src.id", "src"] doesn't occur due to fm.Normalize()
+		f[path] = nil
+	case 2:
+		if m, ok := f[sub[0]]; !ok || m == nil {
+			f[sub[0]] = make(filter)
+		}
+		f[sub[0]].add(sub[1])
+	default:
```

Comment:
what are the cases where this would be empty? is there a reason we do not return an error?

Reply:
changed to use `strings.Cut` instead to make it clear that there are no other options